deploy/docker/cp-edge/sync-routes.py

```diff
@@ -595,88 +759,56 @@ def load_pods_for_runs_with_endpoints():
-                if nginx_sensitive_route_definitions:
-                        for nginx_sensitive_route_definition in nginx_sensitive_route_definitions:
-                                added_route_file.write(nginx_sensitive_route_definition)
-
-        if has_custom_domain:
-                print('Adding {} route to the server block {}'.format(path_to_route, service_hostname))
-                add_custom_domain(service_hostname, path_to_route)
-
-        service_url = SVC_URL_TMPL.format(external_ip=service_hostname,
-                                          edge_location=service_spec["edge_location"] if service_spec["edge_location"] else "",
-                                          edge_port=str(edge_service_port),
-                                          service_name=service_spec["service_name"],
-                                          is_default_endpoint=service_spec["is_default_endpoint"],
-                                          external_schema=edge_service_external_schema)
-        run_id = service_spec["run_id"]
-        if run_id in service_url_dict:
-                service_url = service_url_dict[run_id] + ',' + service_url
-        service_url_dict[run_id] = service_url
+                create_service_location(service_spec, added_route, service_url_dict)
 
-# Once all entries are added to the template - run "nginx -s reload"
+# reload nginx first time to enable all urls with out custom dns with "nginx -s reload"
 # TODO: Add error handling, if non-zero is returned - restore previous state
 if len(routes_to_add) > 0 or len(routes_to_delete) or routes_were_updated:
-        print('Reloading nginx config')
+        print('Reloading nginx config to enable non custom DNS runs')
         check_output('nginx -s reload', shell=True)
 
-
 # For all added entries - call API and set Service URL property for the run:
 # -- Get ServiceExternalIP from the EDGE-labeled service description
 # -- http://{ServiceExternalIP}/{PodID}-{svc-port-N}-{N}
 
 for run_id in service_url_dict:
-        # make array of json objects
-        service_urls_json = '[' + service_url_dict[run_id] + ']'
-        update_svc_method = os.path.join(api_url, API_UPDATE_SVC.format(run_id=run_id))
-        print('Assigning service url ({}) to RunID: {}'.format(service_urls_json, run_id))
+        if run_id not in runs_with_custom_dns:
+                update_svc_url_for_run(run_id)
 
-        data = json.dumps({'serviceUrl': service_urls_json})
-        response_data = call_api(update_svc_method, data=data)
-        if response_data:
-                print('Service url ({}) assigned to RunID: {}'.format(service_urls_json, run_id))
-        else:
-                print('Service url was not assigned due to API errors')
+#Now handle all routes with custom DNS
+for added_route in routes_with_custom_dns:
+        service_spec = services_list[added_route]
+        dns_services_pool.apply_async(create_dns_service_location, (service_spec, added_route, service_url_dict))
```

Comment:
I wonder if it is safe to modify the same `service_url_dict` dictionary in multiple threads including the main one. Correct me if I'm wrong but this `apply_async` call just copies the arguments and therefore the dictionary in multiple threads is just not the same one anymore.